File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n,1),suffix(n,1);
        for(int i=1; i<n; i++) {
            res[i] = res[i-1] * nums[i-1];
        }
        for(int i=n-2; i>=0; i--) {
            suffix[i] = suffix[i+1] * nums[i+1];
            res[i] *= suffix[i];
        }
        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int ans[] = new int[n];
        int suffix[] = new int[n];
        int prefix[] = new int[n];
        prefix[n-1] = 1;
        suffix[0] = 1;
        for(int i=1; i<n; i++) {
            suffix[i] = nums[i-1] * suffix[i-1];
        }
        for(int i=n-2; i>=0; i--) {
            prefix[i] = nums[i+1] * prefix[i+1];
        }
        for(int i=0; i<n; i++) {
            ans[i] = suffix[i] * prefix[i];
        }
        return ans;
    }
}